atl-checker/src/algorithms/certain_zero/search_strategy/linear_programming_search.rs

```diff
@@ -0,0 +1,224 @@
+use crate::game_structure::lcgs::ir::symbol_table::SymbolIdentifier;
+use minilp::{LinearExpr, OptimizationDirection, Problem, Variable};
+use priority_queue::PriorityQueue;
+use std::collections::HashMap;
+use std::option::Option::Some;
+use std::sync::Arc;
+
+/// A SearchStrategyBuilder for building the LinearProgrammingSearch strategy.
+pub struct LinearProgrammingSearchBuilder {
+    pub game: IntermediateLcgs,
+}
+
+impl SearchStrategyBuilder<AtlVertex, LinearProgrammingSearch> for LinearProgrammingSearchBuilder {
+    fn build(&self) -> LinearProgrammingSearch {
+        LinearProgrammingSearch::new(self.game.clone())
+    }
+}
+
+/// Search strategy using ideas from linear programming to order next vertices,
+/// based on distance from the vertex to a region that borders the line between true/false in the formula.
+pub struct LinearProgrammingSearch {
+    /// Priority based on distance, lowest distance highest priority
+    queue: PriorityQueue<Edge<AtlVertex>, i32>,
+    game: IntermediateLcgs,
+    phi_mapper: ConstrainedPhiMapper,
+    /// Maps the hash of a Phi and usize of state, to a result state and distance
+    result_cache: HashMap<AtlVertex, (State, i32)>,
+    /// Maps phi to a LinearConstrainedPhi
+    phi_cache: HashMap<Arc<Phi>, LinearConstrainedPhi>,
+    // TODO - could add a new cache, to compare distance between states and use this to guesstimate distance
+    // TODO - to acceptance region, based on previous results (Mathias supervisor suggestion)
+}
+
+impl LinearProgrammingSearch {
+    pub fn new(game: IntermediateLcgs) -> LinearProgrammingSearch {
+        LinearProgrammingSearch {
+            queue: PriorityQueue::new(),
+            game,
+            phi_mapper: ConstrainedPhiMapper::new(),
+            result_cache: HashMap::new(),
+            phi_cache: HashMap::new(),
+        }
+    }
+}
+
+impl SearchStrategy<AtlVertex> for LinearProgrammingSearch {
+    /// Simply returns the edge with highest priority (i.e lowest distance)
+    fn next(&mut self) -> Option<Edge<AtlVertex>> {
+        self.queue.pop().map(|(edge, _)| edge)
+    }
+
+    /// Takes a Vec of Edges holding ATLVertices, and puts these in the queue,
+    /// based on the calculated distance from its state to acceptance region from formula
+    fn queue_new_edges(&mut self, edges: Vec<Edge<AtlVertex>>) {
+        for edge in edges {
+            let distance = self.get_distance_of_edge(&edge);
+
+            // Add edge and distance to queue
+            if let Some(dist) = distance {
+                self.queue.push(edge, -dist);
+            } else {
+                // Todo what should default value be, if cannot be calculated? For now: first priority
+                self.queue.push(edge, 0);
+            }
+        }
+    }
+}
+
+impl LinearProgrammingSearch {
+    /// Finds the distance of an edge using linear programming
+    fn get_distance_of_edge(&mut self, edge: &Edge<AtlVertex>) -> Option<i32> {
+        let source = edge.source();
```

Comment:
Did you find that source had better performance than targets?

Reply:
The new `2021-05-14_13-29-22-bench-lin-source-dist-bench-lin-ave-target-dist.tar.gz` benchmark results will show us

Reply:
Just skimmed through the report and using source seems to be slightly better. I think the intuition for it is, that
1) In the worst case, you have to explore a big part of the graph anyway, so calculating the average of targets each time is just overhead.
2) In the best and average cases, where we terminate earlier, calculating the distance for the targets, is still overhead, but you are not going to visit many of those targets anyway.
When using the source, we have more edges with the same priority, but only calculate the distance for vertices that we actually visit.

This is speculation, though. Ideally it should be measured in tests.

Reply:
Coolio

Reply:
Thats a good point, maybe we can use that in another paper